version: '3.1'
services:
  redis:
    image: redis
    ports:
      - '6380:6379'

  postgres_db:
    image: postgres:16.0
    env_file:
      - .env
    environment:
      - PGDATA=/var/lib/postgresql/data/pgdata
    ports:
      - '5433:5432'
    volumes:
      - pg_data:/var/lib/postgresql/data/pgdata
    healthcheck:
      test: [ 'CMD-SHELL', 'pg_isready' ]
      interval: 10s
      timeout: 5s
      retries: 5

  courses:
    build: .
    tty: true
    ports:
      - '8000:8000'
    depends_on:
      postgres_db:
        condition: service_healthy
    command: >
      sh -c "python manage.py migrate && 
      python manage.py create_superuser &&
      python manage.py runserver 0.0.0.0:8000"
    env_file:
      - .env
    environment:
      - POSTGRES_HOST=postgres_db

  celery:
    build: .
    tty: true
    command: celery -A config worker -l INFO
    depends_on:
      - redis
      - courses
    environment:
      - CELERY_BROKER_URL=redis://redis:6379/0
      - CELERY_RESULT_BACKEND=redis://redis:6379/0

  celery-beat:
    build: .
    tty: true
    command: celery -A config beat -l INFO -S django
    depends_on:
      - redis
      - courses
    environment:
      - CELERY_BROKER_URL=redis://redis:6379/0
      - CELERY_RESULT_BACKEND=redis://redis:6379/0

volumes:
  pg_data:
